lib/jwt/jwk/rsa.rb

```diff
@@ -23,50 +25,97 @@ def public_key
       end
 
       def kid
-        sequence = OpenSSL::ASN1::Sequence([OpenSSL::ASN1::Integer.new(public_key.n),
-                                            OpenSSL::ASN1::Integer.new(public_key.e)])
-        OpenSSL::Digest::SHA256.hexdigest(sequence.to_der)
+        @kid ||= generate_kid
```

Comment:
This is a little problematic. I would like to do it so the @kid would only be set once. Doing it after `super` would potentially set the kid twice in the initializer.

I would like to do it like this but the keypair that `generate_kid` needs is not initialized yet.
```
def initialize(keypair, kid = nil)
  raise ArgumentError, 'keypair must be of type OpenSSL::PKey::RSA' unless keypair.is_a?(OpenSSL::PKey::RSA)
  super(keypair, kid || generate_kid)
end
```

A litte ugly solution but would work is to have the kid generation as a static method that gets the public_key as a parameter. I think I'll go with that.

Reply:
Did something in [ec9a4c7](https://github.com/jwt/ruby-jwt/pull/375/commits/ec9a4c77947e0270ac00f0391a0ad68d2b4cc832)

Reply:
"Doing it after super would potentially set the kid twice in the initializer.", at the moment I don't understand why, but you have a better overview of the JWK RSA code flow.

Reply:
I meant in cases where the kid is not given. Then the super will set the `@kid` instance variable to `nil` and then the derived class will set `@kid` to the generated kid. 

What i changed this to give the generated kid to the parent if no kid is provided to the derived class, then @kid is only set once in the super.

Reply:
Is it a problem, if kid is set to nil first? I see no problem with that and the code looks much simpler to understand.

Reply:
I feel that im starting to over analyse this, sorry for that :) But when the kid is set from outside the parent it could to be set via attribute writer. Maybe `@kid` should not be used from the derived classes at all but the `kid` accessors instead.

Reply:
Yep! That's true. Then we remove `@kid = kid` from `initialize` method in `KeyAbstract` class after PR are merged? Or can you remove it with this PR please?

Reply:
I was maybe more referring to not setting `@kid` directly from the RSA, EC or HMAC classes but via some accessor. 

What about something like this [6754a58](https://github.com/jwt/ruby-jwt/pull/375/commits/6754a58a024df6abea29cb2f75d1e0b343cb2c36)

Reply:
Very nice! I agree.